Proposing LIFO for connection pooling

TaeMinPark opened the pull request:
### Why
- SQLAlchemy's connection pool is using the Queue data structure to provide the connection pools to threads round-robin basis. Which leads to maintain all connections even though the server's load is not that high enough to maintain all connections.
- Using the Stack data structure could solve this problem as connections that are not working will be disconnected from the database.

### Changes
- Added `StackPool` in `lib/sqlalchemy/pool/impl.py`.
- Added `Stack` in  `lib/sqlalchemy/util/queue.py`.

zzzeek commented:
this appears to just change the popleft() to a pop().    So, LIFO vs. FIFO.  Sorry I'm being dense but am not seeing the advantage?   the order of the QueuePool doesn't actually matter at all, it's a "pool" and could just be an unordered set.   if your pool size is set to 5 but you only use 2 connections at a time, your queue will have only 2 connections.   it won't create a third unless the program asks for three concurrent connections at some point.

Here is a demo:
```

>>> from sqlalchemy import create_engine
>>> e = create_engine("postgresql://scott:tiger@localhost/test")
>>> c1 = e.connect()
>>> e.pool.status()
'Pool size: 5  Connections in pool: 0 Current Overflow: -4 Current Checked out connections: 1'
>>> c2 = e.connect()
>>> e.pool.status()
'Pool size: 5  Connections in pool: 0 Current Overflow: -3 Current Checked out connections: 2'
>>> c2.close()
>>> c1.close()
>>> e.pool.status()
'Pool size: 5  Connections in pool: 2 Current Overflow: -3 Current Checked out connections: 0'
>>> c3 = e.connect()
>>> e.pool.status()
'Pool size: 5  Connections in pool: 1 Current Overflow: -3 Current Checked out connections: 1'
>>> c4 = e.connect()
>>> e.pool.status()
'Pool size: 5  Connections in pool: 0 Current Overflow: -3 Current Checked out connections: 2'
>>> c3.close()
>>> c4.close(
... )
>>> e.pool.status()
'Pool size: 5  Connections in pool: 2 Current Overflow: -3 Current Checked out connections: 0'

```

pool size is five but we only used two at a time.  So there's only two actual connections.    How does LIFO change that ?

I can apply the change like this:

```
diff --git a/lib/sqlalchemy/util/queue.py b/lib/sqlalchemy/util/queue.py
index 1958702c7c..99b92bd2fe 100644
--- a/lib/sqlalchemy/util/queue.py
+++ b/lib/sqlalchemy/util/queue.py
@@ -196,4 +196,4 @@ class Queue:
 
     # Get an item from the queue
     def _get(self):
-        return self.queue.popleft()
+        return self.queue.pop()

```

do the same steps, I get the identical output:

```

>>> from sqlalchemy import create_engine
>>> e = create_engine("postgresql://scott:tiger@localhost/test")
>>> c1 = e.connect()
>>> e.pool.status()
'Pool size: 5  Connections in pool: 0 Current Overflow: -4 Current Checked out connections: 1'
>>> c2 = e.connect()
>>> e.pool.status()
'Pool size: 5  Connections in pool: 0 Current Overflow: -3 Current Checked out connections: 2'
>>> c2.close()
>>> c1.close()
>>> e.pool.status()
'Pool size: 5  Connections in pool: 2 Current Overflow: -3 Current Checked out connections: 0'
>>> c3 = e.connect()
>>> e.pool.status()
'Pool size: 5  Connections in pool: 1 Current Overflow: -3 Current Checked out connections: 1'
>>> c4 = e.connect()
>>> e.pool.status()
'Pool size: 5  Connections in pool: 0 Current Overflow: -3 Current Checked out connections: 2'
>>> c3.close()
>>> c4.close()
>>> e.pool.status()
'Pool size: 5  Connections in pool: 2 Current Overflow: -3 Current Checked out connections: 0'

```

zzzeek commented:
Can you clarify "the DB doesnt have to maintain all connections" ?   Are you referring towards the DB timing out those connections from the server side?    Hence if there are five pooled connections, but only two are in play for a long period of time, the three older connections get timed out and aren't reconnected because they were "first" in.    What database are you targeting and what server-side disconnection mechanism?  haproxy?   I'd like to work up a mock demonstration of this and additionally there's no need to add a new pool, we'd just change Queue to use popleft() and document that it is actually a "stack" nowadays.

zzzeek commented:
additionally this should start with a bug report on bitbucket, rationale, etc.   I really don't like to start with PRs for things like this, sorry.

zzzeek commented:
also, counterpoint.   the app pools five connections, during a low period, only 2 at a time are used.   using FIFO, all 5 of the connections are kept fresh, so if the app has a sudden burst of activity, there isn't a reconnect burden up front.   but this is not as valuable IMO as reducing connections for idle processes.

propose we add a flag to QueuePool use_lifo=True.   we can default it to True and put it in 1.3.

also feel free to evaluate an alternative pool architecture I worked up at https://bitbucket.org/zzzeek/haalchemy/src/master/haalchemy/clients/sqlalchemy/pool.py  .  This pool uses an actively sorted list of connections so you can numerically assign priority to individual connections.   this pool was intended for a more complicated use case and I haven't seen a strong need to pull it in.

TaeMinPark commented:
Plus, I will take an auto-scaling situations as an example. Things are different.

Let's say 10 instances are using 5 connections each. And due to the increase of CPU loads, let's say the number of instances became 20. However, in some cases, instances might not need all 5 connections. CPU load increase could have been driven by jobs that don't use DB. 

There are some special cases that the number of connections is more important than the burden of reconnecting at the sudden burst of activity.

zzzeek commented:
this also needs a flag for create_engine pool_use_lifo over here: https://github.com/zzzeek/sqlalchemy/blob/master/lib/sqlalchemy/engine/strategies.py#L117

TaeMinPark commented:
Changed.

- Added `use_lifo` field to `QueuePool` and `Queue`. Default to False.
- Added descriptions for `use_lifo`.
- Removed `StackPool` and `Stack`.

zzzeek commented:
this looks great, but we need to test it somehow.  i was going to try to write one today but im not sure if I can get to it.

basically, use a Mock for the connections and then count.  here's a demo:

from sqlalchemy import pool
import mock

c1, c2, c3 = mock.Mock(), mock.Mock(), mock.Mock()
connections = [c1, c2, c3]


def creator():
    return connections.pop(0)

p = pool.QueuePool(creator)


pc1 = p.connect()
pc2 = p.connect()
pc3 = p.connect()

pc1.close()
pc2.close()
pc3.close()


def test_queue():
    for i in range(3):
        pc1 = p.connect()
        assert pc1.connection is c1
        pc1.close()

        pc1 = p.connect()
        assert pc1.connection is c2
        pc1.close()

        pc1 = p.connect()
        assert pc1.connection is c3
        pc1.close()


def test_stack():
    for i in range(3):
        pc1 = p.connect()
        assert pc1.connection is c3
        pc1.close()

        pc1 = p.connect()
        assert pc1.connection is c3

        pc2 = p.connect()
        assert pc2.connection is c2
        pc2.close()

        pc1.close()


#test_queue()
test_stack()

TaeMinPark commented:
Yup! I will add a test at `/test/engine/pool.py`.

TaeMinPark commented:
Added `test_stack` at `sqlalchemy.test.engine.test_pool.QueuePoolTest`.

zzzeek commented:
OK I will pull this into gerrit.  i want to also add the test_queue b.c. it is always good to validate what it does normally

zzzeek commented:
Dear contributor -

This pull request is being moved to Gerrit, at https://gerrit.sqlalchemy.org/#/c/zzzeek/sqlalchemy/+/848, where it may be tested and reviewed more closely.  As such, the pull request itself is being marked "closed" or "declined", however your contribution is merely being moved to our central review system. Please register at https://gerrit.sqlalchemy.org#/register/ to send and receive comments regarding this item.

TaeMinPark commented:
Checked!

benjamin-lim commented:
I like this PR

waitforDev commented:
i need this PR

mitrilmad commented:
I love this PR
